### PR TITLE
Add apply view subcommand

### PR DESCRIPTION
**Pull request by shiywang**
implement part of  https://github.com/kubernetes/community/pull/287
@ymqytw @pwittrock @AdoHe

### CODE REVIEW

**Comment by pwittrock** [comment]
> What kind of resouce_printer should we use here to output last-applied-confguration ?

We should support `-o json` and `-o yaml`.  Yaml should be the default.

> Can kubectl apply view last-applied support output multiple configuration of resources at a time?

Yes I think that makes sense.  The resource builder should already support this if you just copy how create works.

File: pkg/kubectl/cmd/apply_view.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/kubernetes/pkg/api/annotations"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+var (
+	applyViewLong = templates.LongDesc(`
+		Get the last-applied-configuration annotations by type/name or file.
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: Use dedent to take care of the indentation.  See other commands.

File: pkg/kubectl/cmd/apply_view.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+		You can use --output jsonpath={...} to extract specific values using a jsonpath expression.`)
+
+	applyViewExample = templates.Examples(`
+		# Get the last-applied-configuration annotations by type/name
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: I'd reword this to be more user oriented.

> # Get the last applied configuration file for a resource by type/name.
> # This is retrieved from the last-applied-configuration annotation on the resource

File: pkg/kubectl/cmd/apply.go [hunk 3]
```diff
@@ -128,6 +128,10 @@ func NewCmdApply(f cmdutil.Factory, out, errOut io.Writer) *cobra.Command {
 	cmdutil.AddPrinterFlags(cmd)
 	cmdutil.AddRecordFlag(cmd)
 	cmdutil.AddInclude3rdPartyFlags(cmd)
+
+	// apply subcommands
+	cmd.AddCommand(NewCmdApplyView(f, out))
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply.go]: We are going to want to add a `last-applied` subcommand.  And then add a `view` subcommand to that.  It will make the `-h` work.

File: pkg/kubectl/cmd/apply_view.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+	usage := "that contains the last-applied-configuration annotations"
+	cmdutil.AddFilenameOptionFlags(cmd, &options.FilenameOptions, usage)
+	return cmd
+}
+
+func ApplyViewLastApplied(f cmdutil.Factory, errOut io.Writer, cmd *cobra.Command, args []string, options *ApplyOptions) error {
+	if args[0] != lastApplied {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: Not necessary if you make a `last-applied` sub-command, and a `view` sub-sub-command

File: pkg/kubectl/cmd/apply_view.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+		return err
+	}
+
+	annotationMap, err := info.Mapping.MetadataAccessor.Annotations(info.Object)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := annotationMap[annotations.LastAppliedConfigAnnotation]; !ok {
+		fmt.Fprintf(errOut, warningNoLastAppliedConfigAnnotation)
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: We should probably just return an error.

return fmt.Errorf("No last applied configuration found for resource %s/%s.  This will happen if the resource was not created with apply", kind, name)

File: pkg/kubectl/cmd/apply_view.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+	infos, err := r.Infos()
+	if err != nil {
+		return err
+	}
+	info := infos[0]
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: We should visit all of the matching resources.  See how `apply` does things for an example

File: pkg/kubectl/cmd/apply_view.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	o.printer, _, err = kubectl.GetPrinter(outputFormat, templateFile, false, cmdutil.GetFlagBool(cmd, "allow-missing-template-keys"))
+	if err != nil {
+		return err
+	}
+	printer, generic, err := cmdutil.PrinterForCommand(cmd)
+	if err != nil {
+		return err
+	}*/
+
+	fmt.Println(annotationMap[annotations.LastAppliedConfigAnnotation])
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: See other commands for example on how to use a printer to use the `-o` flag

File: pkg/kubectl/cmd/apply_view.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	if args[0] != lastApplied {
+		return cmdutil.UsageError(cmd, "for now apply %s only support last-applied")
+	}
+
+	mapper, typer, err := f.UnstructuredObject()
+	if err != nil {
+		return err
+	}
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply_view.go]: Is this the right way to do this?  We want to respect the namespace flag if set.

**Comment by pwittrock** [comment]
FYI @jessfraz

**Comment by pwittrock** [comment]
@jessfraz Probably makes sense for you to review and for @AdoHe to approve.

File: pkg/kubectl/cmd/apply_view.go (hunk 8 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: if we explicit flag `-n` or `--namespace`. then the `DefaultNamspace()` will return the value we set to `cmdNamespace`, sorry, I don't quit understand what do you mean by respect the namespace flag if set ?
btw, do we need to also support `all-namespace` and `selector`?

File: pkg/kubectl/cmd/apply_view.go (hunk 2 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: updated

File: pkg/kubectl/cmd/apply.go (hunk 3 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply.go]: done

File: pkg/kubectl/cmd/apply_view.go (hunk 4 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: ok, already change to subcommand

File: pkg/kubectl/cmd/apply_view.go (hunk 5 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: copy that

File: pkg/kubectl/cmd/apply_view.go (hunk 6 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: yeah, change to `visit`, just not sure if it's multiple resources before

File: pkg/kubectl/cmd/apply_view.go (hunk 7 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: ok, now I'm using `runtime.Decode` to convert json to an obj, seems codecs have some problem, the yaml output file did not contain some of information. also some of information should be filtered.

Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: I'm not sure if i'm doing the right way, do we have a exist command provide some convenient way to convert a json string to a printable object ?

Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: current output : https://gist.github.com/shiywang/e5409a69dfe0ad7d73779dbbceaeb026

**Comment by shiywang** [comment]
current output : https://gist.github.com/shiywang/caca7c2ddc8537aa6b94c75d022ef348

File: pkg/kubectl/cmd/apply_view.go (hunk 7 above)
Comment (shiywang) [inline pkg/kubectl/cmd/apply_view.go]: update, ptal

**Comment by shiywang** [comment]
@jessfraz @AdoHe ptal, thanks

**Comment by shiywang** [comment]
this branch has been messed up by mistake, will open another pr
